lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -1175,11 +1099,20 @@ void ReabstractionInfo::specializeConcreteAndGenericSubstitutions(
   {
     SmallVector<Substitution, 4> List;
 
+    // Define a substitution map for re-mapping interface types of
+    // the specialized function to contextual types of the caller.
+    auto SpecializedInterfaceToCallerArchetypeMap =
+        SpecializedGenericSig->getSubstitutionMap(
+            [&](SubstitutableType *type) -> Type {
+              return SpecializedInterfaceToCallerArchetypeMapping.lookup(
+                  type->getCanonicalType());
+            },
+            LookUpConformanceInSignature(*SpecializedGenericSig));
+
     SpecializedGenericSig->getSubstitutions(
         [&](SubstitutableType *type) -> Type {
           auto Replacement =
-              SpecializedInterfaceToCallerArchetypeMapping.lookup(
-                  type->getCanonicalType());
+              Type(type).subst(SpecializedInterfaceToCallerArchetypeMap);
           if (Replacement)
             return Replacement;
```

Comment:
Is it possible for this to fail still? If not, this whole part can just become SpecializedGenericSig->getSubstitutions(SpecializedInterfaceToCallerArchetypeMap, List);